src/sa.py

```diff
@@ -1,15 +1,149 @@
 import argparse
+import fasta, fastq
+import re
 
+def getSuffixes(x: memoryview):
+    """
+    Gets all suffixes from a string
+    """
+    suffixes = [x[i:] for i in range(0, len(x))] 
+    return suffixes
+
+def getTrailingNumber(s):
+    m = re.search(r'\d+$', s)
+    return (s[:m.start()], int(s[m.start():]))
+
+def strcmp(s1, index, x):
+    max_i = min(len(s1), len(x)-index)
+    i = 0
+    while i < max_i and s1[i] == x[index+i]:
+        i += 1
+        if i >= max_i:
+            return 0
+    if s1[i] > x[index+i]: return 1
+    else: return -1
+
+
+def search(sa, pattern, genome):
+    low = 0
+    high = len(sa)-1
+
+    if low == high:
+        cmp = strcmp(pattern, sa[low], genome)
+        if cmp == 0:
+            yield sa[low] + 1
+            return
+
+    lowCompared = False
+    highCompared = False
+    while True:
+        mid = (high+low)//2
+        midSuffix = sa[mid]
+        cmp = strcmp(pattern, midSuffix, genome)
+
+        if cmp==0:
+            while cmp==0 and len(pattern) > len(genome) - midSuffix:
+                mid += 1
+                if mid >= len(sa): return
```

Comment:
If you return here, you won't `yield` anything more. But `midSuffix` was a match (because `cmp == 0`), so shouldn't you at least yield that?

I could be wrong, but it looks to me like the inner `while` runs upwards in the suffix array to find the end of the region that matches, and that is fine, but if you reach the end, you terminate the generator without emitting anything. What you want to do is put the end point at `len(sa) - 1` (or `len(sa)` if you had excluded the end point from the interval, the way Odin intended).